File: sdk/fsharp/src/HiTechCloudSDK/CloudVm.fs
namespace HiTechCloud.SDK

/// Cloud Virtual Machine: upgrade, resources, templates, ISO, boot, PXE
type CloudVmResource(httpClient: HttpClient) =

    /// GET /api/service/{id}/upgrade - List resource upgrade options
    member this.GetUpgradeOptions() =
        async {
            return! httpClient.getAsync("/api/service/{id}/upgrade")
        }

    /// POST /api/service/{id}/upgrade - Estimate or request upgrade
    member this.RequestUpgrade() =
        async {
            return! httpClient.postAsync("/api/service/{id}/upgrade")
        }

    /// GET /api/service/{id}/resources - Show available and used resources
    member this.GetResources() =
        async {
            return! httpClient.getAsync("/api/service/{id}/resources")
        }

    /// GET /api/service/{id}/templates/{vmid} - List rebuild templates
    member this.ListRebuildTemplates() =
        async {
            return! httpClient.getAsync("/api/service/{id}/templates/{vmid}")
        }

    /// GET /api/service/{id}/vms/{vmid}/usage/memory - Memory usage graph
    member this.GetMemoryUsage() =
        async {
            return! httpClient.getAsync("/api/service/{id}/vms/{vmid}/usage/memory")
        }

    /// GET /api/service/{id}/vms/{vmid}/images - List ISO images for VM
    member this.ListIsoImages() =
        async {
            return! httpClient.getAsync("/api/service/{id}/vms/{vmid}/images")
        }

    /// POST /api/service/{id}/vms/{vmid}/images - Mount/Unmount ISO image
    member this.MountIso() =
        async {
            return! httpClient.postAsync("/api/service/{id}/vms/{vmid}/images")
        }

    /// POST /api/service/{id}/vms/{vmid}/boot - Change boot order
    member this.SetBootOrder() =
        async {
            return! httpClient.postAsync("/api/service/{id}/vms/{vmid}/boot")
        }

    /// POST /api/service/{id}/vms/{vmid}/tuntap - Enable PXE
    member this.EnablePxe() =
        async {
            return! httpClient.postAsync("/api/service/{id}/vms/{vmid}/tuntap")
        }

File: sdk/clojure/src/hitechcloud/sdk/cloud_vm.clj
(ns hitechcloud.sdk.cloud_vm
  (:require [hitechcloud.sdk.http-client :as http]))

;; Cloud Virtual Machine: upgrade, resources, templates, ISO, boot, PXE

(defn get_upgrade_options "GET /api/service/{id}/upgrade - List resource upgrade options" [client & [params]] (http-get client "/api/service/{id}/upgrade" params))

(defn request_upgrade "POST /api/service/{id}/upgrade - Estimate or request upgrade" [client & [params]] (http-post client "/api/service/{id}/upgrade" params))

(defn get_resources "GET /api/service/{id}/resources - Show available and used resources" [client & [params]] (http-get client "/api/service/{id}/resources" params))

(defn list_rebuild_templates "GET /api/service/{id}/templates/{vmid} - List rebuild templates" [client & [params]] (http-get client "/api/service/{id}/templates/{vmid}" params))

(defn get_memory_usage "GET /api/service/{id}/vms/{vmid}/usage/memory - Memory usage graph" [client & [params]] (http-get client "/api/service/{id}/vms/{vmid}/usage/memory" params))

(defn list_iso_images "GET /api/service/{id}/vms/{vmid}/images - List ISO images for VM" [client & [params]] (http-get client "/api/service/{id}/vms/{vmid}/images" params))

(defn mount_iso "POST /api/service/{id}/vms/{vmid}/images - Mount/Unmount ISO image" [client & [params]] (http-post client "/api/service/{id}/vms/{vmid}/images" params))

(defn set_boot_order "POST /api/service/{id}/vms/{vmid}/boot - Change boot order" [client & [params]] (http-post client "/api/service/{id}/vms/{vmid}/boot" params))

(defn enable_pxe "POST /api/service/{id}/vms/{vmid}/tuntap - Enable PXE" [client & [params]] (http-post client "/api/service/{id}/vms/{vmid}/tuntap" params))

